west: runners: openocd: fix broken rtt implementation

cfriedt opened the pull request:
* cmake: flash: update cmake to support rtt target
* west: runners: jlink: print RTT server port
* west: runners: openocd: mitigate pylint R0201 warning
* west: runners: openocd: fix rtt implementation

Fixes #81156

Testing done:

```shell
west build -p -S rtt-console -b <board> -t flash samples/hello_world/
...
west rtt -r openocd --cmd-pre-init 'adapter serial <serial>'
...
*** Booting Zephyr OS build v4.0.0-371-g8b7d3d663704 ***
Hello World! <board>
```

```shell
west build -p -S rtt-console -b <board> -t flash samples/hello_world/
...
west rtt -r jlink
...
*** Booting Zephyr OS build v4.0.0-371-g8b7d3d663704 ***
Hello World! <board>
```

> [!NOTE]  
> Due to the init strategy for the RTT buffer, either a power-cycle or reflash is required between `west rtt` invocations
> to observe the Hello World message again, since buffer contents and positions are preserved by default (e.g. from
> mcuboot). A simple reset via gdb is not sufficient to clear the RTT buffer. Alternatively, manually erase the RTT block
> via gdb memory write commands.

Commit:
cmake: flash: update cmake to support rtt target

Add support for the rtt target so that users can run

`west build -p auto -b <board> -t rtt <app>`

similarly to the way that users can now do so with the debug
target, since the rtt target is supposed to be used in a similar
way.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
west: runners: update the rtt implementation for openocd

Previously, rtt start would always fail because the target
(i.e. firmware) had not been started.

```
Info : rtt: Searching for control block 'SEGGER RTT'
Info : rtt: No control block found
```

When the command is 'rtt', run the binary via gdb before calling
'rtt start'. Firmware calls `SEGGER_RTT_Init()` shortly after
init which allows OpenOCD to find the RTT control block.

Similarly, only start the 'rtt server' after calling
'rtt start' to avoid any potential race conditions
internally within OpenOCD.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Some questions atm

* Does the JLink runner always load the binary again, or does it simply used what was previously flashed to the device?
  * if the latter case, then it's much easier both for west and twister
* Does the JLink runner run `JLinkRTTClient` in the foreground or background?
  * I assumed it was in the foreground based on the existence of `run_telnet_client()` in `jlink.py`
* Does the JLink runner run the gdb server in the foreground or background?
  * based on `jlink.py`, I believe that the gdb server should be run in the background (which would mean changes to `openocd.py`) 

I'll have to double-check on a different machine and modify as necessary for consistency.

Still need to look at simplifying the Twister integration, but it should really be something along the lines of 

```shell
twister \
  -i \
  -p <board> \
  --device-testing \
  --west-flash \
  --west-runner $RUNNER \
  --device-serial-pty scripts/rtt.sh \
  -xCONFIG_USE_SEGGER_RTT=y \
  -xCONFIG_RTT_CONSOLE=y \
  -T samples/hello_world
```

Kind of awkward, but the `--device-serial-pty` script part might need .elf awareness for openocd, unless we want to use the base of SRAM and the size of SRAM as the search parameters for `rtt setup`. 

In any case, the openocd setup seems to be much more reliable than the JLink setup, although it's possible that the main problem with the JLink setup was due to #81194.

dkalowsk reviewed (dismissed):
haven't tested these.  Overall the changes match to what the commits talk about and seem logical to me.  Thanks for the PR!

rruuaanng commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -369,16 +368,52 @@ def do_attach_debug(self, command, **kwargs):
         self.require(gdb_cmd[0])
         self.print_gdbserver_message()
-        self.run_server_and_client(server_cmd, gdb_cmd)
+        if command == 'rtt':
+            self.print_rttserver_message()
```
It seems that it can be combined with `407 line`.

pdgendt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -29,6 +27,19 @@
+
+def to_num(number):
+    dev_match = re.search(r"^\d*\+dev", number)
+    dev_version = not dev_match is None
```
Since you're touching this

```suggestion
    dev_version = dev_match is not None
```

cfriedt replied:
Fixed!

pdgendt replied:
Not anymore

fabiobaltieri commented on cmake/flash/CMakeLists.txt:
```diff
@@ -168,6 +168,8 @@ foreach(target flash debug debugserver attach)
   elseif(target STREQUAL attach)
     set(comment "Debugging ${BOARD}")
+  elseif(target STREQUAL rtt)
+    set(comment "Debugging ${BOARD}")
```
should this have something other than debugging? maybe just `RTT ${BOARD}`

cfriedt replied:
Fixed!

Commit:
west: runners: jlink: print RTT server port

Similar to how print_gdbserver_message() prints GDB server info
when "west debug" is run, print RTT server info when "west rtt"
is run.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Just to follow up here

> 1. Does the JLink runner always load the binary again, or does it simply used what was previously flashed to the device? If the latter case, then it's much easier both for west and twister

JLink does not load the binary via the `rtt` runner, and so that functionality was removed for `openocd` runner as well to be consistent.

> 2. Does the JLink runner run `JLinkRTTClient` in the foreground or background? I assumed it was in the foreground based on the existence of `run_telnet_client()` in `jlink.py`

It is run in the foreground.

> 3. Does the JLink runner run the gdb server in the foreground or background? Based on `jlink.py`, I believe that the gdb server should be run in the background (which would mean changes to `openocd.py`)

It is run in the background (or rather, the TCP port is left open for gdb clients to connect).

The two runners should behave quite similarly now.

cfriedt commented:
This is in a good enough place for now. I'll likely follow-up shortly with at twister PR - hopefully it will be acceptible.

cfriedt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -369,16 +368,51 @@ def do_attach_debug(self, command, **kwargs):
+            if rtt_address is None:
+                raise ValueError("RTT Control block not be found")
+
+            # gdb is not intended to be used interactively with rtt. It is only used to load the
+            # binary, so we cannot prompt the user to press return to see more text. Disable
+            # pagination for rtt.
```
Woops - this comment needs to be reworded. It's not accurate anymore

cfriedt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -369,16 +368,51 @@ def do_attach_debug(self, command, **kwargs):
+            self.print_rttserver_message()
+            server_proc = self.popen_ignore_int(server_cmd)
+            try:
+                # gdb is still necessary to load, setup rtt, etc, but it wll be detached and will
```
This comment needs to be updated (or maybe removed?)

cfriedt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -369,16 +368,51 @@ def do_attach_debug(self, command, **kwargs):
+            self.run_server_and_client(server_cmd, gdb_cmd)
+        elif command == 'rtt':
+            self.print_rttserver_message()
+            server_proc = self.popen_ignore_int(server_cmd)
```
It would probably make sense to include this call in the `try` / `except` block below.

cfriedt replied:
Fixed!

cfriedt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -369,16 +368,51 @@ def do_attach_debug(self, command, **kwargs):
+            self.print_rttserver_message()
+            server_proc = self.popen_ignore_int(server_cmd)
+            try:
+                # gdb is still necessary to load, setup rtt, etc, but it wll be detached and will
+                # quit so that rtt client can take over.
```
This comment needs to be updated (or removed?)

cfriedt replied:
Fixed!

cfriedt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -29,6 +27,19 @@
 DEFAULT_OPENOCD_TARGET_HANDLE = "_TARGETNAME"
 
+
+def to_num(number):
+    dev_match = re.search(r"^\d*\+dev", number)
+    dev_version = dev_match is not None
+
+    num_match = re.search(r"^\d*", number)
+    num = int(num_match.group(0))
+
+    if dev_version:
+        num += 1
+
+    return num
```
I might put this change into a separate commit, since it's really just addressing a warning about R0201

Commit:
west: runners: openocd: mitigate pylint R0201 warning

Previously, there was a warning that the to_num() method can
(and probably should) be unbound from OpenOcdBinaryRunner.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* fixed comments
* split fix for R0201 into separate commit

cfriedt commented:
* reworded git commits to be be more uniform

pdgendt commented:
@topisani PTAL

fabiobaltieri commented:
reassigning, does not really fall under build system, or any other area that I can tell

cfriedt commented:
Looks like unrelated tests are failing - rebased (hopefully the unrelated issues are fixed already)

cfriedt commented:
@pdgendt - please revisit when you have a moment

topisani commented:
Sorry for the wait, LGTM
